### PR TITLE
feat: enable custom prefixes/triples for RDF API

### PR DESCRIPTION
**Pull request by svandenhoek**
What are the main changes you did:
- Prefixes are now user-defined (per database).
- Addition of custom triples (note that these will be added to EVERY API call for that database)
- Default is no prefixes except namespace of database itself (can be changed by adding `settings: _settings/rdf_api` to templates).
  - Alternative is adding an extra advanced setting in `_SYSTEM_` for global prefixes/RDF.
- If invalid RDF is added, settings will be stored but API will fail.

In case multiple database are called, the prefixes/custom triples well be merged together.

how to test:
- Create a new database of "FAIR_DATA_HUB" and change the values in the advanced setting "custom_rdf". Any changes should immediately be present in the API.

todo:
- [ ] updated docs in case of new feature
- [ ] added/updated tests
- [ ] added/updated testplan to include a test for this fix, including ref to bug using # notation

### CODE REVIEW

**Review by mswertz — CHANGES_REQUESTED** [review]
nice. Can you please adapt such that in case this setting is not set the old situation exits?

**Comment by svandenhoek** [comment]
> nice. Can you please adapt such that in case this setting is not set the old situation exits?

Done. Old behaviour occurs if any of the schemas does not have a `custom_rdf` setting. Maybe future implementation could be a `_SYSTEM` setting `custom_rdf` which contains the fallback (or even a default that always gets added anyway).

File: backend/molgenis-emx2-rdf/src/main/java/org/molgenis/emx2/rdf/RDFService.java [hunk 1]
```diff
@@ -148,24 +153,29 @@ public void describeAsRDF(
       final Schema... schemas) {
     try {
       final ModelBuilder builder = new ModelBuilder();
-      builder.setNamespace("rdf", NAMESPACE_RDF);
```
Comment (mswertz) [inline backend/molgenis-emx2-rdf/src/main/java/org/molgenis/emx2/rdf/RDFService.java:151]: What happens if same namespace is defined twice?
Could we simply keep this and then add the RDF setting to override?

File: backend/molgenis-emx2-rdf/src/main/java/org/molgenis/emx2/rdf/RDFService.java [hunk 2]
```diff
@@ -148,24 +153,29 @@ public void describeAsRDF(
...
-      builder.setNamespace("fdp-o", NAMESPACE_ORG);
 
+      // Defines if all used schemas have a custom_rdf setting.
+      boolean allIncludeCustomRdf = true;
       // Define the schemas at the start of the document.
       for (final Schema schema : schemas) {
         final Namespace ns = getSchemaNamespace(schema);
         builder.setNamespace(ns);
+        // Adds custom RDF to model.
+        if (schema.hasSetting(SETTING_CUSTOM_RDF)) {
```
Comment (mswertz) [inline backend/molgenis-emx2-rdf/src/main/java/org/molgenis/emx2/rdf/RDFService.java:164]: hmm, per schema means they all add up when multiple schemas? that is then also no problem

File: e2e/README.md [hunk 3]
```diff
@@ -1,26 +1,43 @@
 We have e2e tests using playwright
```
Comment (mswertz) [inline e2e/README.md:1]: these changes are unrelated? so should be in other PR?

File: e2e/playwright.config.ts [hunk 4]
```diff
@@ -39,6 +39,15 @@ export default defineConfig({
     {
       name: 'chromium',
       use: { ...devices['Desktop Chrome'] },
+      testIgnore: '*/admin!*.spec.ts'
```
Comment (mswertz) [inline e2e/playwright.config.ts:42]: and this also unrelated?

File: e2e/tests/api/rdf/admin!custom-rdf-edit.spec.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
+const { chromium } = require('playwright');
```
Comment (mswertz) [inline e2e/tests/api/rdf/admin!custom-rdf-edit.spec.ts:1]: ah, that is why you made the e2e changes.

I would not want to use playwright for API tests. Instead use Java see for examples WebApiSmokeTests

**Review by mswertz — CHANGES_REQUESTED** [review]
please don't use e2e for API tests. 
e2e is brittle and expensive. Instead use Java see WebApiSmokeTest.testRdfApi for example on how that looks like
It uses RestAssured as the framework.

Also, could you use a simpler data model possibly?
Fair data hub creation is very expensive

File: backend/molgenis-emx2-rdf/src/main/java/org/molgenis/emx2/rdf/RDFService.java (hunk 1 above)
Comment (svandenhoek) [reply backend/molgenis-emx2-rdf/src/main/java/org/molgenis/emx2/rdf/RDFService.java:151]: I'll make sure to add tests for:
* Identical copies -> should be non-issue
* Same namespace with different prefix -> no clue tbh
* Same prefix different namespace -> probably replaces and last-added is used?

File: backend/molgenis-emx2-rdf/src/main/java/org/molgenis/emx2/rdf/RDFService.java (hunk 2 above)
Comment (svandenhoek) [reply backend/molgenis-emx2-rdf/src/main/java/org/molgenis/emx2/rdf/RDFService.java:164]: Exactly. Current behavior simply combines all `custom_rdf` data if using multiple schemes. Special cases:
* If one or more has no `custom_rdf`, default list is added (as a scheme assumes these prefixes are present)
* If schemes have an empty `custom_rdf` instead of non-existing, this should cause the default addition to be skipped.

File: e2e/tests/api/rdf/admin!custom-rdf-edit.spec.ts (hunk 5 above)
Comment (svandenhoek) [reply e2e/tests/api/rdf/admin!custom-rdf-edit.spec.ts:1]: Noted. Yeah Playwright had some issues as in VS code it worked but local cmd/CI failed. Will remove code here though create separate PR in case someone in the future needs logged-in e2e tests (with example test).

**Comment by svandenhoek** [comment]
e2e changes removed here and available in #4054

**Comment by svandenhoek** [comment]
Failing test labelled by CircleCI as [FLAKY](https://circleci.com/blog/introducing-test-insights-with-flaky-test-detection/).

**Comment by mswertz** [comment]
> Details

You mean this one? 
image

I restarted the test to see...

**Comment by svandenhoek** [comment]
> > Details
> 
> You mean this one? image
> 
> I restarted the test to see...

Indeed. Also, when I ran that test through IDEA it did not fail. So no clue why it fails.

Scherm­afbeelding 2024-07-23 om 16 54 46

(sidenote: `/apps/aggregates/src/views/view-dashboard.vue` was edited by google formatter, but doubt that can be the cause)

**Comment by mswertz** [comment]
Test does fail on my machine. Might it by you have other data?

On my machine it fails on this line: 
image

**Review by mswertz — APPROVED** [review]
